examples/demo-unused-variable-warner.rkt

```diff
@@ -0,0 +1,106 @@
+#lang racket
+(require "../main.rkt"
+         (for-syntax nanopass/base
+                     syntax/name
+                     syntax/id-set)
+         racket/base)
+
+(provide (except-out (all-from-out racket/base) #%module-begin))
+
+(begin-for-syntax
+  (define (demo prog)
+    (define defined-variables (make-hasheqv))
+    (define used-variables (make-hasheqv))
+    
+    (define (mark-defined! idt phase)
+      (unless (and (eq? '_ (syntax-e idt))
+                   (syntax-original? (syntax-local-introduce idt)))
+        (unless (hash-ref defined-variables phase #f)
+          (hash-set! defined-variables phase (mutable-free-id-set #:phase phase)))
+        (free-id-set-add! (hash-ref defined-variables phase)
+                          idt)
+        ))
+    (define (mark-used! idt phase)
+      (unless (hash-ref used-variables phase #f)
+        (hash-set! used-variables phase (mutable-free-id-set #:phase phase)))
+      (free-id-set-add! (hash-ref used-variables phase)
+                        idt))
+    
+    (define (collect-ids! ids phase)
+      (syntax-case ids ()
+        [a (identifier? #'a) (mark-defined! #'a phase)]
+        [() (void)]
+        [(a . b) (mark-defined! #'a phase)
+                 (collect-ids! #'b phase)]
+        ))
+    
+    (define-pass mark-unused-variable : (FullyExpandedProgram ModuleBeginForm) (prog) -> (FullyExpandedProgram ModuleBeginForm) ()
+      (ModuleBeginForm : ModuleBeginForm (prog phase) -> ModuleBeginForm ()
+                       [(#%plain-module-begin ,s ,[ml* 0 -> ml*] ... ,[ml 0 -> ml])
+                        prog])
+      (Expr : Expr (prog phase) -> Expr ()
+            [,x 
+             (mark-used! x phase)
+             prog]
+            [(#%top ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%variable-reference ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%variable-reference-top ,s ,x)
+             (mark-used! x phase)
+             prog]
+            [(#%plain-lambda ,s ,fml ,[body*] ... ,[body])
+             (collect-ids! fml phase)
+             prog]
+            [(case-lambda ,s [,fml ,[body*] ... ,[body]] ...)
+             (for ([f fml])
+               (collect-ids! f phase))
```

Comment:
using `in-list` and `in-hash` for `for` macros to avoid constructing and iterating over a stream